src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -126,7 +126,7 @@ private void OnCompilationEventsGenerated_NoLock(ImmutableArray<CompilationEvent
                     if (HasActionsForEvent(compilationEvent, actionCounts))
                     {
                         _pooledEventsWithAnyActionsSet.Add(compilationEvent);
-                        analyzerState.OnCompilationEventGenerated(compilationEvent, actionCounts);
+                        await analyzerState.OnCompilationEventGeneratedAsync(compilationEvent, actionCounts, cancellationToken).ConfigureAwait(false);
```

Comment:
either it is guaranteed that _analyzerStateMap doesnt change here or you need to have local copy of _analyzerStateMap (just var map = _analyzerStateMap) since it is immutable collection